File: backend/src/agents/analyst.py
import asyncio
from src.agents.base import BaseAgent, AgentProfile, TaskResult
from src.integrations.kalibr_router import routed_completion
from typing import Optional

class AnalystAgent(BaseAgent):
    def __init__(self):
        super().__init__(AgentProfile(
            agent_id="analyst",
            name="Analyst",
            role="analyst",
            description="Processes data, generates insights and recommendations. Routes LLM calls through Kalibr.",
        ))

    async def execute(self, task: dict) -> TaskResult:
        description = task.get("description", "")
        context = task.get("context", [])

        await self.emit("analyst.processing", {
            "message": f"Analyzing: {description[:100]}...",
        })
        await asyncio.sleep(0.5)

        context_str = "\n".join([f"- {c['from_agent']}: {c['data']}" for c in context]) if context else "No prior context."

        system_prompt = f"""You are an expert financial analyst specializing in DeFi and on-chain economics. Provide a thorough analysis with actionable recommendations.

Context from other agents:
{context_str}

Structure your response as:
## Analysis Report
### Key Insights (3-5 bullet points with specific numbers and comparisons)
### Risk Assessment (LOW/MEDIUM/HIGH with detailed reasoning for each risk category: protocol, market, IL, liquidity)
### Recommendations (specific, actionable steps with exact parameters — pool ranges, allocation percentages, rebalancing triggers)
### Confidence Level (percentage with reasoning based on data quality and market conditions)

Be decisive. Give specific numbers, not ranges. State your recommendation clearly."""

        analysis = await routed_completion(
            goal="analysis",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": description},
            ],
        )

        await self.emit("analyst.insight", {
            "message": "Analysis complete",
            "summary": analysis[:200],
        })

        return TaskResult(success=True, data={"analysis": analysis}, cost=0.002)

    async def bid(self, task: dict) -> Optional[float]:
        if task.get("type") == "analysis":
            return 0.002
        return None

File: frontend/src/lib/agents/analyst.ts
import { BaseAgent } from "./base";
import { TaskResult, TaskInput } from "./types";
import { llmCompletion } from "../llm";

export class AnalystAgent extends BaseAgent {
  constructor() {
    super({
      agentId: "analyst",
      name: "Analyst",
      role: "analyst",
      description: "Processes data, generates insights and recommendations.",
    });
  }

  async execute(task: TaskInput): Promise<TaskResult> {
    const description = task.description || "";
    const context = task.context || [];

    this.emit("analyst.processing", {
      message: `Analyzing: ${description.slice(0, 100)}...`,
    });

    const contextStr = context.length
      ? context.map((c) => `- ${c.from_agent}: ${c.data}`).join("\n")
      : "No prior context.";

    const systemPrompt = `You are an expert financial analyst specializing in DeFi and on-chain economics. Provide a thorough analysis with actionable recommendations.

Context from other agents:
${contextStr}

Structure your response as:
## Analysis Report
### Key Insights (3-5 bullet points with specific numbers and comparisons)
### Risk Assessment (LOW/MEDIUM/HIGH with detailed reasoning for each risk category: protocol, market, IL, liquidity)
### Recommendations (specific, actionable steps with exact parameters — pool ranges, allocation percentages, rebalancing triggers)
### Confidence Level (percentage with reasoning based on data quality and market conditions)

Be decisive. Give specific numbers, not ranges. State your recommendation clearly.`;

    const analysis = await llmCompletion("analysis", [
      { role: "system", content: systemPrompt },
      { role: "user", content: description },
    ]);

    this.emit("analyst.insight", {
      message: "Analysis complete",
      summary: analysis.slice(0, 200),
    });

    return { success: true, data: { analysis }, cost: 0.002 };
  }

  async bid(task: TaskInput): Promise<number | null> {
    return task.type === "analysis" ? 0.002 : null;
  }
}
